crates/bevy_ecs/src/entity/visit_entities.rs

```diff
@@ -1,48 +1,75 @@
+use std::iter;
+
 use crate::{
     entity::Entity,
     identifier::masks::{IdentifierMask, HIGH_MASK},
     world::World,
 };
+pub use bevy_ecs_macros::IterEntities;
 
 use super::EntityHashMap;
 
-/// Operation to map all contained [`Entity`] fields in a type to new values.
-///
-/// As entity IDs are valid only for the [`World`] they're sourced from, using [`Entity`]
-/// as references in components copied from another world will be invalid. This trait
-/// allows defining custom mappings for these references via [`EntityMappers`](EntityMapper), which
-/// inject the entity mapping strategy between your `MapEntities` type and the current world
-/// (usually by using an [`EntityHashMap<Entity>`] between source entities and entities in the
-/// current world).
+/// Apply an operation to all entities in a container.
 ///
-/// Implementing this trait correctly is required for properly loading components
-/// with entity references from scenes.
-///
-/// ## Example
+/// This is implemented by default for types that implement [`IterEntities`] or
+/// where `&T` and `&mut T` both implement [`IntoIterator`].
 ///
-/// ```
-/// use bevy_ecs::prelude::*;
-/// use bevy_ecs::entity::MapEntities;
+/// It may be useful to implement directly for types that can't produce an
+/// iterator for lifetime reasons, such as those involving internal mutexes.
+pub trait VisitEntities {
+    /// Apply an operation to all contained entities.
+    fn visit_entities<F: FnMut(Entity)>(&self, f: F);
+    /// Apply an operation to mutable references to all entities.
+    fn visit_entities_mut<F: FnMut(&mut Entity)>(&mut self, f: F);
+}
+
+impl<T> VisitEntities for T
+where
+    T: IterEntities,
+{
+    fn visit_entities<F: FnMut(Entity)>(&self, f: F) {
+        self.iter_entities().for_each(f);
+    }
+    fn visit_entities_mut<F: FnMut(&mut Entity)>(&mut self, f: F) {
+        self.iter_entities_mut().for_each(f);
+    }
+}
+
+/// Produce an iterator over all contained entities.
 ///
-/// #[derive(Component)]
-/// struct Spring {
-///     a: Entity,
-///     b: Entity,
-/// }
+/// This is implemented by default for types  where `&T` and `&mut T` both implement
+/// [`IntoIterator`].
 ///
-/// impl MapEntities for Spring {
-///     fn map_entities<M: EntityMapper>(&mut self, entity_mapper: &mut M) {
-///         self.a = entity_mapper.map_entity(self.a);
-///         self.b = entity_mapper.map_entity(self.b);
-///     }
-/// }
-/// ```
-pub trait MapEntities {
-    /// Updates all [`Entity`] references stored inside using `entity_mapper`.
-    ///
-    /// Implementors should look up any and all [`Entity`] values stored within `self` and
-    /// update them to the mapped values via `entity_mapper`.
-    fn map_entities<M: EntityMapper>(&mut self, entity_mapper: &mut M);
+/// It may be useful to implement directly for types that can't produce an
```

Comment:
It's also useful for `Box` because you can't have a type that returns `impl Trait`.

Reply:
Hmm, how do you mean?

Reply:
You can't have `Box<dyn IterEntities>`, Rust just won't let you. I think this is the main reason. Maybe worth to mention.

Reply:
Oh, VisitEntities isn't object-safe either since its methods have generics. That can be changed with a slight hit to callsite ergonomics, i.e. `c.visit_entities(|ent| println!("{ent}"))` would have to be `c.visit_entities(&mut |ent| println!("{ent}"))`

Reply:
Ah, I forgot that `VisitEntities` also not object-safe (I thought about `IterEntities`). Need to think about it...

Reply:
To be fair, `MapEntities` isn't object-safe either :shrug:

Reply:
Ah, right... I forgot that it accepts entity mapper. Doesn't matter then :)